src/coredump/coredump.c

```diff
@@ -810,49 +827,51 @@ static int submit_coredump(
         return 0;
 }
 
-static void map_context_fields(const struct iovec *iovec, const char* context[]) {
-
-        static const char * const context_field_names[] = {
-                [CONTEXT_PID] = "COREDUMP_PID=",
-                [CONTEXT_UID] = "COREDUMP_UID=",
-                [CONTEXT_GID] = "COREDUMP_GID=",
-                [CONTEXT_SIGNAL] = "COREDUMP_SIGNAL=",
-                [CONTEXT_TIMESTAMP] = "COREDUMP_TIMESTAMP=",
-                [CONTEXT_RLIMIT] = "COREDUMP_RLIMIT=",
-                [CONTEXT_HOSTNAME] = "COREDUMP_HOSTNAME=",
-                [CONTEXT_COMM] = "COREDUMP_COMM=",
-                [CONTEXT_EXE] = "COREDUMP_EXE=",
-        };
+static int cache_process_info(ProcessInfoCache *cache, const struct iovec_wrapper *iovw) {
+        unsigned n, i, count = 0;
+        char *unit;
+        int r;
 
-        unsigned i;
+        assert(cache);
+        assert(iovw);
+        assert(iovw->count >= _META_ARGV_MAX);
 
-        assert(iovec);
-        assert(context);
+        /* The cache does not allocate any memory on its own */
 
-        for (i = 0; i < ELEMENTSOF(context_field_names); i++) {
-                char *p;
+        memzero(cache, sizeof(ProcessInfoCache));
 
-                if (!context_field_names[i])
-                        continue;
+        for (n = 0; n < iovw->count; n++) {
+                struct iovec *iovec = iovw->iovec + n;
 
-                p = memory_startswith(iovec->iov_base, iovec->iov_len, context_field_names[i]);
-                if (!p)
-                        continue;
+                for (i = 0; i < ELEMENTSOF(meta_field_names); i++) {
+                        char *p;
 
-                /* Note that these strings are NUL terminated, because we made sure that a trailing NUL byte is in the
-                 * buffer, though not included in the iov_len count. (see below) */
-                context[i] = p;
-                break;
+                        p = startswith(iovec->iov_base, meta_field_names[i]);
```

Comment:
because process_kernel and process_backtrace use it to store strings only (done by gather_pid_metadata_*). If the iovecs are received from the socket (process_socket) we make sure that the iovecs are null terminated, see the comment in process_socket()

Reply:
please add a comment about that. And maybe, use memory_startswith() here, plus add an assert that checks that the buffers are NUL terminated to document this behaviour in code

Reply:
done.